Add docs describing the process of writing a new schema (#375)
spacetelescope/rad: docs/creating.rst

highlights all the important considerations for RAD schema creation.

Before you begin
----------------

Before you start writing your schema, you should have a clear idea of what data
you wish to store in a Roman file and how you want to organize it. Remember that
RAD supports a hierarchical data model, so you can topically organize your data
under headings and subheadings (or even deeper) as you see fit. In particular,
you should have a clear idea of the following:

    #. The name of the schema you wish to use. Be sure to determine the version
       number of the schema. Typically, if the schema is new, it will be
       ``1.0.0``. So the schema name will be ``<name>-1.0.0.yaml``.

    #. Where in RAD you wish to locate your schema. For example if it is for a
       reference file then it should be located in the ``reference_files``
       directory.

    #. The keywords for your fields and their hierarchical organization.

    #. The data types of all the fields that you wish to store. In particular,
       you need to pay attention to the following:

        * Which fields will be primitive data types like ``int``, ``float``,

>           ``str``, or ``bool``.

Reviewer: Should this mention these are Python primitive types since the types have different names in YAML. Or may be include the YAML names in parenthesis.

.. note::

    All external tags should end with a ``-<major version>.*`` version
    specifier. Rather than a specific version number, this is a wildcard that
    will match any version of that tag. This is to ensure that the schema is not
    tied to a specific version of the external tag.

Create the Schema Boilerplate
-----------------------------

After you have created your new schema file in the location under the name you
have chosen, you should add the following to your blank file

.. code:: yaml

    YAML 1.1
    ---
    $schema: asdf://stsci.edu/datamodels/roman/schemas/rad_schema-1.0.0
    id: asdf://stsci.edu/datamodels/roman/schemas/<file name of schema>  # No .yaml

    title: <Title of the schema>
    description: |
        <A long description of the schema>

    ...

The ``YAML 1.1`` needs to be on the very first line of the file, while the
``...`` needs to be on the second to last line of the file with the final line
being completely empty.

Add Your Fields
---------------

Now we will populate your schema with the fields you wish to use. In almost all
cases you will want to use an ``object`` type for your top level of the schema,

> for other cases see :ref:`alternate-fields`.  n this case your add the following

Reviewer: ```suggestion
for other cases see :ref:`alternate-fields`.  In this case you add the following
```

after your ``description`` in the boilerplate:

.. code:: yaml

    type: object
    properties:
        <first keyword>:
            title: <Title of the field>
            description: |

>                 <A long description of the field>

Reviewer: ```suggestion
                <A long description of the field, can be multiline>
```

You will repeat this step for each of the top-level fields you wish to add.

Populate a Field's Sub-Schema
*****************************

After the field's ``description`` at the same indentation level as the
``description`` keyword, you will start to add the sub-schema for the field.
There are several different possibilities at this point:

* Primitive type.
    Things like ``int``, ``float``, ``str``, or ``bool``. In this case you will
    add the following:

    .. code:: yaml

        type: <type>

.. note::

    The ``<type>`` for a Python float is ``number`` and the ``<type>`` for a

>     Python bool is ``boolean``.

Reviewer: OK, so may be here is the proper place to add the YAML `integer` and `string` names.

    tag. If you create an object or another array you likewise add the metadata
    in the same way as if it were a top-level field only indented appropriately.

Special Field Considerations
****************************

There are a few special considerations that you might need to take into account
when creating your schema:

* Enum.
    If you have a field that can only take on a specific set of values, you can
    use the ``enum`` keyword to specify the possible values. For example:

    .. code:: yaml

        enum: [<value1>, <value2>, <value3>]

* Multiple Possibilities.
    If a field can take on multiple different types, you can use the ``oneOf``
    combiner to specify the different possibilities. For example:

    .. code:: yaml

        oneOf:
          - type: <type1>
          - type: <type2>
          - type: <type3>

    where further metadata can be added to each of the types as needed.

    .. note::

        Sometimes you might want to have a field which is required, but which
        may not take on any values at all. In this case you can use the
        ``null`` type as one of the possibilities in the ``oneOf`` combiner.

> 

Reviewer: Is it worth mentioning `anyOf`?

Author: Currently we are only using the `oneOf` combiner.
